core/src/appleMain/kotlin/CBPeripheralCoreBluetoothPeripheral.kt

```diff
@@ -378,9 +387,42 @@ internal class CBPeripheralCoreBluetoothPeripheral(
         }
 
         val platformDescriptor = discoveredServices.obtain(descriptor)
-        return connection.execute<DidUpdateValueForDescriptor> {
+        val updatedDescriptor = connection.execute<DidUpdateValueForDescriptor> {
             centralManager.read(cbPeripheral, platformDescriptor)
-        }.descriptor.value as NSData
+        }.descriptor
+
+        // CBDescriptor "value" property type depends on cbDescriptor.UUID.UUIDString
+        // https://stackoverflow.com/questions/52816049/how-to-convert-cbdescriptor-value-to-string
+        // https://developer.apple.com/documentation/corebluetooth/characteristic-descriptors
+        when (updatedDescriptor.UUID.UUIDString) {
+            CBUUIDCharacteristicFormatString -> {
+                return updatedDescriptor.value as NSData
+            }
+            CBUUIDCharacteristicAggregateFormatString,
+            CBUUIDCharacteristicUserDescriptionString,
+            -> {
+                return (updatedDescriptor.value as String)
+                    .encodeToByteArray()
+                    .toNSData()
+            }
+            CBUUIDCharacteristicExtendedPropertiesString,
+            CBUUIDClientCharacteristicConfigurationString,
+            CBUUIDServerCharacteristicConfigurationString,
+            CBUUIDL2CAPPSMCharacteristicString,
```

Comment:
Yes Apple API is awkward, indeed.
 I've also searched online a bit and the only useful doc I've found is [this one](https://www.bluetooth.com/wp-content/uploads/Files/Specification/HTML/Core-54/out/en/host/generic-attribute-profile--gatt-.html), but still is not clear to me which is that correct type we should expect for each different value.

`Do you have a real-world example to test this one on, as that would be ideal, otherwise I suppose we go with our best-guess/stack-overflow comments?`

i've tested with an IOT device that mount an ESP32 BLE Board on it. 
I'm using Kable to rewrite in Kotlin Multiplatform this [Espressif SDK](https://github.com/espressif/esp-idf-provisioning-android) which reads descriptors and use it for different kind of messages.

I've also tested with a Xiaomi Mi Smart Band 5 (nothing work related, its just that i have it with me), but since is not something I need for work, I've just verified that the code didn't crash.

This week I'm kind of busy at work, but maybe in the weekend I can run some tests on these two devices and provide you some extra information about this.

Reply:
Thanks for the detail. I agree with your deductions in the other comment as well. The only reference I found in the docs is for the type alias https://developer.apple.com/documentation/corebluetooth/cbl2cappsm/ which also hints that this the value for `CBUUIDL2CAPPSMCharacteristicString` is exposed as a raw 16bit integer, and is not a `NSNumber` or `NSData` object. It is for `L2CAP` so to test a real world example would require reading characteristics from a device that supports `L2CAP` channels. I'll update this thread if I find a concrete example.